src/lib/common/payload_uart.cpp

```diff
@@ -28,6 +28,99 @@ static bool _useLineBuffer = true;
+static SemaphoreHandle_t _postTxSemaphore = nullptr;
+
+// Forward declaration of the post-transmission callback function
+static void pluartPostTransactionCb(SerialHandle_t *handle);
+
+// Set transaction hardware control functions
+void enableTransactions(HardwareControlFunc preTxFunc, HardwareControlFunc postTxFunc) {
+  // Assign the transaction start and end callbacks
+  _preTxFunction = preTxFunc;
+  _postTxFunction = postTxFunc;
+  // Assign the post-transaction callback function,
+  //  which wraps _postTxFunction in some additional checks to ensure transaction is complete.
+  uart_handle.postTxCb = pluartPostTransactionCb;
+}
+
+// Start a Tx transaction
+// TODO - return false if transaction already started?
+void startTransaction() {
+  taskENTER_CRITICAL();
+  _transactionInProgress = true;
+  _writeDuringTransaction = false;
+  taskEXIT_CRITICAL();
+  if (_preTxFunction) {
+    _preTxFunction(); // External start transaction function. Eg, enable Tx for Half-Duplex
+  }
+}
+
+/** End a Tx transaction, called by user.
+// Blocks until all PLUART writes are complete then fires _postTxFunction
+// Description: First safely check if there's a transaction to end, safely wait for all Tx writing
+                to finish, then _postTxFunction().
+TODO - If users have multiple tasks interacting with PLUART transactions,
+       this will heavily bork. This is not a currently supported use case.
+ **/
+bool endTransaction(uint32_t wait_ms) {
+  taskENTER_CRITICAL();
+  // If not in a transaction, bail and tell the user.
+  if (!_transactionInProgress) {
+    taskEXIT_CRITICAL();
+    return false;
+  }
+
+  // If there were no writes during the transaction, don't wait for write complete semaphore token.
+  if (!_writeDuringTransaction) {
+    _transactionInProgress = false;
+    taskEXIT_CRITICAL();
+    _postTxFunction(); // Safely call the function
+    return true;
+  }
+
+  // transaction is in progress, and need to call _postTxFunction when writing is complete.
+  _transactionInProgress = false;
+  taskEXIT_CRITICAL();
+
+  // Exit ciritcal while we wait for final write to complete.
+  // The first while check will fail if data has not yet been added to stream buffer,
+  //    So need to make sure we block on first semaphore take first.
+  // After each transmission compeltes, TC interrupt will call pluartPostTransactionCb,
+  //    which will give a token to the semaphore.
+  // Each subsequent write inside the transaction will take from the semaphore,
+  //    so only the final write will leave a token available here.
+  bool ret_value = true;
+  if (xSemaphoreTake(_postTxSemaphore, pdMS_TO_TICKS(wait_ms)) == pdFALSE) {
+    // Failed to take sempaphore within transaction timeout
+    ret_value = false; // TODO - maybe better to assert(false) here?
+  }
+  // Call the post-transaction function after the final transmission is fully complete
+  if (_postTxFunction != nullptr) _postTxFunction();
```

Comment:
I like this a lot better, I can understand wanting to offload it into its own task, but this gives a nice way to give the user control of when this callback will fire. Creative and intuitive.

Reply:
👍